cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll

```diff
@@ -0,0 +1,326 @@
+    (input.isParameter(0) or input.isParameter(1)) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * A prefix `operator*` member function for an iterator type.
+ */
+class IteratorPointerDereferenceMemberOperator extends MemberFunction, TaintFunction {
+  IteratorPointerDereferenceMemberOperator() {
+    this.hasName("operator*") and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * An `operator++` or `operator--` member function for an iterator type.
+ */
+class IteratorCrementMemberOperator extends MemberFunction, DataFlowFunction, TaintFunction {
+  IteratorCrementMemberOperator() {
+    (
+      this.hasName("operator++") or
+      this.hasName("operator--")
+    ) and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierAddress() and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValueDeref()
+  }
+}
+
+/**
+ * A member `operator->` function for an iterator type.
+ */
+class IteratorFieldMemberOperator extends Operator, TaintFunction {
+  IteratorFieldMemberOperator() {
+    this.hasName("operator->") and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierAddress() and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * An `operator+` or `operator-` member function of an iterator class.
+ */
+class IteratorBinaryArithmeticMemberOperator extends MemberFunction, TaintFunction {
+  IteratorBinaryArithmeticMemberOperator() {
+    (this.hasName("operator+") or this.hasName("operator-")) and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      input.isQualifierObject() or
+      input.isParameter(0)
+    ) and
+    output.isReturnValue()
+  }
+}
+
+/**
+ * An `operator+=` or `operator-=` member function of an iterator class.
+ */
+class IteratorAssignArithmeticMemberOperator extends MemberFunction, DataFlowFunction, TaintFunction {
+  IteratorAssignArithmeticMemberOperator() {
+    (
+      this.hasName("operator+=") or
+      this.hasName("operator-=")
+    ) and
+    this.getDeclaringType() instanceof Iterator
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    input.isQualifierAddress() and
+    output.isReturnValue()
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
+    output.isQualifierObject()
+    or
+    input.isQualifierObject() and
+    output.isReturnValueDeref()
```

Comment:
Now that https://github.com/github/codeql/pull/4125/commits/c083c6235debcf041c2373e89a98b41ec6d60e92 is merged, it's appropriate to add taint or data flow from `isReturnValueDeref` to `isQualifierObject` for all functions that return `this`.